src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectSystem/RazorProjectHost.cs

```diff
@@ -0,0 +1,182 @@
+        }
+
+        // Should only be called from the UI thread.
+        private ProjectSnapshotManagerBase GetProjectManager()
+        {
+            _commonServices.ThreadingService.VerifyOnUIThread();
+
+            if (_projectManager == null)
+            {
+                _projectManager = (ProjectSnapshotManagerBase)_workspace.Services.GetLanguageServices(RazorLanguage.Name).GetRequiredService<ProjectSnapshotManager>();
+            }
+
+            return _projectManager;
+        }
+
+        protected override Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            // Don't try to evaluate any properties here since the project is still loading and we require access
+            // to the UI thread to push our updates.
+            //
+            // Just subscribe and handle the notification later.
+            var receiver = new ActionBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>>(OnProjectChanged);
+            _subscription = _commonServices.ActiveConfiguredProjectSubscription.JointRuleSource.SourceBlock.LinkTo(
+                receiver,
+                initialDataAsNew: true,
+                suppressVersionOnlyUpdates: true,
+                ruleNames: new string[] { RazorGeneral.SchemaName, });
+
+            return Task.CompletedTask;
+        }
+
+        protected override async Task DisposeCoreAsync(bool initialized)
+        {
+            _subscription.Dispose();
+
+            await ExecuteWithLock(async () =>
+            {
+                if (_current != null)
+                {
+                    await _commonServices.ThreadingService.SwitchToUIThread();
+
+                    var projectManager = GetProjectManager();
+                    projectManager.HostProjectRemoved(_current);
+                    _current = null;
+                }
+            });
+        }
+
+        // Internal for testing
+        internal async Task OnProjectChanged(IProjectVersionedValue<IProjectSubscriptionUpdate> update)
+        {
+            await ExecuteWithLock(async () =>
+            {
+                if (IsDisposing || IsDisposed)
+                {
+                    return;
+                }
+
+                await _commonServices.ThreadingService.SwitchToUIThread();
+                var projectManager = GetProjectManager();
+
+                var razorProperties = await _commonServices.ActiveConfiguredProjectRazorProperties.GetRazorGeneralPropertiesAsync().ConfigureAwait(false);
+                var languageVersion = await razorProperties.RazorLangVersion.GetEvaluatedValueAtEndAsync().ConfigureAwait(false);
+
+                if (string.IsNullOrEmpty(languageVersion))
+                {
+                    languageVersion = "2.1";
+                }
+
+                var hostProject = new HostProject(_commonServices.UnconfiguredProject.FullPath, languageVersion);
+                if (_current == null)
+                {
+                    _current = hostProject;
```

Comment:
so i'm a little confused about the lifetimes here. This class is exported as a MEF service which makes me believe there's 1 per IDE yet we have a _current which represents the current host project which has a file path. 

I thought `HostProject` represented VS' version of csproj were there can be > 1? I think I'm missing something 😄

Reply:
Nah, CPS works differently. 

You get one of these per project in the solution.

Reply:
> You get one of these per project in the solution.

ah beautiful, ok!